rfcs/0022-kviklet.md

```diff
@@ -0,0 +1,86 @@
+# Kviklet 
+
+- Feature Name: `kviklet`
+- Start Date: 2024-03-29
+- RFC PR: [mbta/technology-docs/#0022](https://github.com/mbta/technology-docs/pull/22)
+- Asana task: [Kviklet (run reviewed DB queries on prod)](https://app.asana.com/0/1200506724882024/1206873943983582/f)
+- Status: Proposed
+
+## Summary
+
+Set up [Kviklet](https://kviklet.dev/), a self-hosted tool to allow teams access to the production database to run one-off queries, after review by other team members, similar to code review.
+
+The status-quo and alternative is that teams can only run database queries by deploying code to prod.
+
+## Motivation
+
+I need the ability to run arbitrary read queries against the Glides prod database, in order to:
+- Debug issues in prod. Examples:
+  - [Is a field that we think is unique actually unique in practice?](https://app.asana.com/0/1200273269966439/1205996566957450)
+  - Gather quantitative data about realtime tracking problems, instead of just ad hoc reports when somebody notices them.
+  - Sign in isn't working for only one operator. Do we have their info in the database?
+- Run Analytics. Examples:
+  - [What are people putting into free-text comments, and does it suggest any features we should build for them?](https://app.asana.com/0/616151179860796/1206468619394322)
+  - How long before a trip are inspectors entering departure times? (This affects prediction quality, and has been a major effort of Glides that we've been iterating on for a while, but it's harder when it's hard to export data.)
+
+This is not just a problem for Glides. Any app with a database is likely to run into similar problems.
+
+This has been a problem for us for a long time, which we haven't found a good solution for yet.
+
+## Status Quo / Alternative
+
+The prod database can only be accessed by the application, so in order to run a query, the Glides team must deploy a code change to the server. This is pretty high overhead that must be paid even for a small one-off query that takes 30 seconds to plan and run locally.
+
+Currently, the Glides app has an admin UI with a way to run a few hand-written queries. Adding or changing these is not easy, because it requires updaing UI and API endpoints in addition to writing the query itself.
+ 
+When [I asked other teams](https://app.asana.com/0/1200506724882024/1206647196651141/f), the recommended approach was to build a system like T-Alerts, where they still have to deploy code with a hardcoded approved query, but have a better system set up to reduce the effort and overhead in writing new queries. Glides was planning on doing this approach until we learned of Kviklet.
+
+Other options were considered and rejected in the asana task linked above.
```

Comment:
I've read the Asana task now and I find I'm still unclear on what alternatives were considered and why they were rejected. The only thing I can pick out is the possibility of setting up a read replica for each database and giving engineers unrestricted access to those, and there was some discussion about the infrastructure cost and the potential benefit for avoiding performance impacts on production. Was there anything else?

I'm also not seeing reference to any other off-the-shelf tools that might exist, and how they would compare to Kviklet — is it the only one of its kind?

Reply:
K, I'll add them.